keps/sig-auth/2799-reduction-of-secret-based-service-account-token/README.md

```diff
@@ -0,0 +1,428 @@
+change obviates the need for auto-generation of secret-based service account
+tokens which are [less secure than the bound token](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#background).
+
+### Goals
+
+- No auto-generation of secret-based service account token.
+- Removal of unused auto-generated secret-based service account tokens
+
+### Non-Goals
+
+- Removal of [explicitly requested secret-based service account tokens](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-service-account-api-token).
+
+## Proposal
+
+- Change the service account control loop in Token Controller to not create
+  secret for service accounts. At the same time, warn usage of auto-created
+  secret-based service account tokens and encourage users to use TokenRequest
+  API or manually-created secret-based service account tokens.
+- Purge auto-generated secret-based service account tokens.
+
+### Notes/Constraints/Caveats
+
+- A warning mechanism should be implemented to help users
+  migrate.
+- Auto generated secret-based service account tokens are those requested by
+  Token Controller and the names have prefix "{service account name}-token-".
+- Only clean up auto-generated tokens which:
+  - are not referenced by pods
+  - have not been used to authenticate for some duration (time duration or number of releases)
+- To consult active usage of secret-based tokens, metric
+  `serviceaccount_legacy_tokens_total` or audit annotation
+  `authentication.k8s.io/legacy-token` could be used.
+
+### Risks and Mitigations
+
+- When feature LegacyServiceAccountTokenNoAutoGeneration is Beta, consumers
+  depending directly on waiting for and reading tokens out of auto-generated
+  secrets might stop working. To mitigate,
+  1. When Alpha, announce the stopping auto-generating token secrets will happen
+     in Beta.
+  2. Emit warnings when using auto-generated token secrets.
+  3. Publish pointers to TokenRequest or the manual secret request flow.
+- When LegacyServiceAccountTokenCleanUp is Beta, usage of auto-generated
+  secret-based token might stop working. To mitigate,
+  1. When Alpha, annouce the cleanup starts at Beta
+  2. Emit warnings when using auto-generated token secrets.
+  3. Add pointers of TokenRequest API and manually created tokens in the validation
+     result.
+
+## Design Details
+
+### Token Controller
+
+#### LegacyServiceAccountTokenNoAutoGeneration:
+
+1. Token Controller stops to create secret for service accounts.
+2. To facilitate LegacyServiceAccountTokenCleanUp, when this feature
+   is Alpha where pod-based usage of auto-generated tokens that are injected
+   by admission controller should have dried up (BoundServiceAccountTokenVolume
+   is GAed in 1.22),
+   - Annotate auto-generated secrets with a date indicating when we started
+     tracking use (e.g."secret.kubernetes.io/used-since: \$date", default to "unknown").
```

Comment:
> it clears last-used (if set) on secrets when they are used

we don't need to do this. this could be problematic as well in case envelop encryption is used. when determining the time that a secret was used:
1. `last-used` if exists and after `tracked-since`
2. defaults to `tracked-since`

we can s/LegacyServiceAccountTokenCleanUp/LegacyServiceAccountTokenNoAutoGeneration to fast forward.